### PR TITLE
Add LLM access for Anthropic endpoints

### PR DESCRIPTION
**Pull request by gabrielcorado**
Related to [RFD 0038e - LLM Access](https://github.com/gravitational/teleport.e/blob/rfd/0038e-llm-access/rfd/0038e-llm-access.md).

Adds LLM access to Anthropic format for the providers: Anthropic API and AWS Bedrock.

>[!NOTE]
> Some changes were not included in this PR, given the increased scope (those changes will be done in separate PRs):
> 1. Usage tracking. The LLM request audit event includes usage fields for input and output tokens. In this PR, this data won't be included.
> 2. Memory budgeting for app services when serving LLM access. Included on the RFD that the app service must have better memory management when serving LLM access, given their size. On this PR, this is not taken into account.

For reviewers, the PR includes 3 main parts of the LLM access:
1. Processing and executing LLM access requests (mainly implemented at `lib/srv/app/llm/anthropic.go`). As mentioned on the RFD, we're relying on the Anthropic SDK to issue requests. This covers API formatings. However, we still need to inspect parts of the request to serve it and generate audit events.
2. Error formatting (`lib/srv/app/llm/errors.go` and `lib/srv/app/llm/anthropic_errors.go`). Errors from providers are not directly returned to Teleport clients. Instead, they're parsed and formatted into stable messages (the original messages are still logged). This is done to improve UX and avoid exposing internal details of the request being served.
3. Testing. To ensure proper use of the SDK and address some edge cases, we've added tests for most flows and cases related to Anthropic access.

Misc:
- Per RFD analysis, the binary increase with the addition of the Anthropic SDK should be ~1.9MB.

### Test Cases
- [x] Access using `curl` and `claude` (provider: `anthropic`) 
  - [x] Mapped model.
  - [x] Streaming requests.
  - [x] Fallback model.
  - [x] Invalid messages/errors returned are from Teleport.
  - [x] Audit events are included in the app session recording. `tsh play -f json <chunk-id>`
  - [x] Prometheus metrics generated.
  
- [x] Access using `curl` and `claude` (provider: `bedrock`) 
  - [x] Mapped model.
  - [x] Streaming requests.
  - [x] Fallback model.
  - [x] Invalid messages/errors returned are from Teleport.
  - [x] Audit events are included in the app session recording. `tsh play -f json <chunk-id>`
  - [x] Prometheus metrics generated.

### CODE REVIEW

File: lib/srv/app/llm/anthropic.go [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+		// Log `err` to get the root cause, not the user-friendly message.
+		log.WarnContext(r.Context(), "anthropic request failed", "error", err)
+		return
+	}
+
+	err = h.executeAnthropicRequest(r.Context(), llm, req, w, r, body)
+	if err != nil {
+		// Here we unwrap the error to get the root cause, not the user-friendly
+		// message.
+		log.WarnContext(r.Context(), "anthropic request failed", "error", errors.Unwrap(trace.Unwrap(err)))
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/srv/app/llm/anthropic.go:80]: This is expected and done this way so that administrators troubleshooting issues can read the provider errors.

File: lib/srv/app/llm/handler.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// readLimitedRequestBody reads request body respecting the app service memory
+// limits.
+func (h *Handler) readLimitedRequestBody(r *http.Request, maxSize int64) (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+	if r.ContentLength > maxSize {
+		return nil, trace.BadParameter("request content is too large, consider splitting the request")
+	}
+	if _, err := buf.ReadFrom(io.LimitReader(r.Body, maxSize+1)); err != nil {
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/srv/app/llm/handler.go:120]: Per the PR description, this is out of scope for this PR and will be addressed in another PR.

File: lib/events/emitter.go [hunk 3]
```diff
@@ -269,7 +269,7 @@ func (l *LoggingEmitter) EmitAuditEvent(ctx context.Context, event apievents.Aud
 	}
 
 	switch event.GetType() {
-	case ResizeEvent, SessionDiskEvent, SessionPrintEvent, AppSessionRequestEvent, "":
+	case ResizeEvent, SessionDiskEvent, SessionPrintEvent, AppSessionRequestEvent, AppSessionLLMRequestSuccessEvent, AppSessionLLMRequestFailureEvent, "":
```
Comment (tigrato) [inline lib/events/emitter.go:272]: is there a special reason to avoid logging LLM audit events besides the cardinality?

File: lib/srv/app/common/audit.go [hunk 4]
```diff
@@ -46,6 +46,8 @@ type Audit interface {
 	OnRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, status uint32, re *AWSResolvedEndpoint) error
 	// OnDynamoDBRequest is called when app request for a DynamoDB API is sent and a response is received.
 	OnDynamoDBRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, status uint32, re *AWSResolvedEndpoint) error
+	// OnLLMRequest is called when app request for LLM inference endpoint is sent and a response is received.
+	OnLLMRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, llmReq LLMRequest, llmResp LLMResponse) error
```
Comment (tigrato) [inline lib/srv/app/common/audit.go:50]: I believe this is fine for now but for my Q2 work, I will need to heavily modify this approach and capture the request and response payloads.

File: lib/srv/app/common/audit.go [hunk 5]
```diff
@@ -235,8 +265,10 @@ func (a *audit) EmitEvent(ctx context.Context, e apievents.AuditEvent) error {
 	recErr := a.cfg.Recorder.RecordEvent(ctx, preparedEvent)
 	event := preparedEvent.GetAuditEvent()
 	var emitErr error
-	// AppSessionRequest events should only go to session recording
-	if event.GetType() != events.AppSessionRequestEvent {
+	// Request and LLM request events should only go to session recording
```
Comment (tigrato) [inline lib/srv/app/common/audit.go:268]: If they should only go to the session recording, instead of calling EmitEvent which calls the audit log and the record event for the session recording, just call the record event. Adding these excludes sounds a bit problematic if everyone remembers using this approach

File: lib/srv/app/llm/handler.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	req.URL.Path = "/v1" + path
+	return next(req)
+}
+
+const (
+	// llmMaxRequestSize represents the max size (in bytes) of the LLM requests.
+	// For LLMs, we use a higher value than `MaxHTTPRequestSize` since those
+	// requests are naturally large.
+	llmMaxRequestSize = 32 * 1024 * 1024
```
Comment (tigrato) [inline lib/srv/app/llm/handler.go:148]: should we reduce this? 32MB per request seems a lot, specially how we handle them

File: lib/srv/app/llm/metrics.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+			//
+			// 1KiB ... 32MiB
+			Buckets: prometheus.ExponentialBuckets(1024, 2, 16),
+		},
+		[]string{teleport.ComponentLabel, "format", "provider", "streaming"},
+	)
+)
+
+func init() {
+	metrics.RegisterPrometheusCollectors(
```
Comment (tigrato) [inline lib/srv/app/llm/metrics.go:83]: don't use the global registry. Pass the process registry and register them manually per https://github.com/gravitational/teleport/pull/51139

File: lib/srv/app/llm/anthropic.go [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+	start := time.Now()
+	defer func() {
+		requestDurationHist.WithLabelValues(
+			teleport.ComponentLLM,
+			llm.Format,
+			llm.Provider,
+		).Observe(time.Since(start).Seconds())
+	}()
+
+	req, body, err := h.prepareAnthropicRequest(llm, r)
```
Comment (tigrato) [inline lib/srv/app/llm/anthropic.go:64]: is there a way we don't need to read the full body before starting the request? If the request is for a streaming endpoint it can add a lot of latency

File: lib/events/emitter.go (hunk 3 above)
Comment (gabrielcorado) [reply lib/events/emitter.go:272]: Two main reasons:
- Like regular app requests, LLM requests can be spammy, especially when using a tool like Claude Code.
- The plan is to expand this event to track the models' actual input/output, which will make it very large.

File: lib/srv/app/common/audit.go (hunk 4 above)
Comment (gabrielcorado) [reply lib/srv/app/common/audit.go:50]: Do you mean how we capture them in the common package here, or how we extract this info from the actual LLM requests?

Comment (tigrato) [reply lib/srv/app/common/audit.go:50]: I believe both

File: lib/srv/app/llm/anthropic.go (hunk 8 above)
Comment (gabrielcorado) [reply lib/srv/app/llm/anthropic.go:64]: It's tricky to start the request without knowing the `model` (required for audit and model mapping) and `max_tokens` (required to determine the timeout and reject long, non-streaming requests), and both can appear anywhere in the request JSON.

In addition, we're treating this request as untrusted because requestors could attempt to expose sensitive information about the provider. That's why we're limiting available endpoints, filtering errors, and ensuring the JSON doesn't contain duplicate keys.

The added latency here is the "price we're paying" for those enforcements. For such endpoints (which usually take seconds to minutes to complete), this increase seems less noticeable than in a regular app access.

File: lib/srv/app/llm/anthropic.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+	if err != nil {
+		apiErr := respondAnthropicError(w, err)
+		auditErr := sessionCtx.Audit.OnLLMRequest(h.closeContext, sessionCtx, r, req, common.LLMResponse{Error: apiErr})
+		if auditErr != nil {
+			log.ErrorContext(r.Context(), "failed to emit audit event", "error", auditErr)
+		}
+		// Log `err` to get the root cause, not the user-friendly message.
+		log.WarnContext(r.Context(), "anthropic request failed", "error", err)
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:72]: nit: should this be a debug log?

File: lib/srv/app/llm/anthropic.go [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+	}
+	auditReq.Model = resolvedModel
+
+	// This function applies Anthropic logic on calculating the adequate request
+	// timeout and also can indicate that the request should be using streaming
+	// instead (due to the potential long response time).
+	timeout := defaultStreamingRequestTimeout
+	if !auditReq.Streaming {
+		var calcErr error
+		timeout, calcErr = anthropic.CalculateNonStreamingTimeout(int(auditReq.MaxTokens), resolvedModel, nil)
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:132]: 10 minutes is still a long time for a http request 😢

File: lib/srv/app/llm/anthropic.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+		evt := dec.Event()
+		if evt.Type == "error" {
+			// Forward errors so clients can handle them mid-response.
+			apiErr := convertAnthropicErrorEvent(evt.Data)
+			n, _ := writeAndFlush(ssestream.Event{
+				Type: "error",
+				Data: anthropicErrorBody(apiErr),
+			})
+			written += n
+			return written, apiErr
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:262]: we are making the assumption that error will always be the last event. just to make sure that's true.

File: lib/srv/app/llm/anthropic.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+		// behavior as standard SSE, convert such error to *apiError.
+		return written, trace.Wrap(convertAnthropicError(err))
+	}
+
+	return written, nil
+}
+
+// writeAnthropicSSEEvent writes an SSE event using standard format (accepted by
+// Anthropic clients).
+func writeAnthropicSSEEvent(w io.Writer, evt ssestream.Event) (int64, error) {
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:286]: would be nice to avoid implementing this ourself.

File: lib/srv/app/llm/anthropic.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+	var (
+		err             error
+		res             = &anthropicRequest{}
+		fieldsOccurency = make(map[string]int)
+	)
+
+	gjson.ParseBytes(buf).ForEach(func(k, v gjson.Result) bool {
+		switch k.String() {
+		case modelRequestKey:
+			if fieldsOccurency[modelRequestKey] == 1 {
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:350]: nit: use a set? or if counting, maybe check the number outside of gjson.ParseBytes.

(otoh, i believe most libraries implement last key wins)

File: lib/srv/app/llm/anthropic.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+	// Messages endpoint.
+	//
+	// https://platform.claude.com/docs/en/api/messages/create
+	"/messages": {},
+}
+
+// List of API request/response JSON keys. Some Anthopic API keys are not
+// defined as constants on the SDK, for those we define them as plain strings.
+//
+// Ref: https://platform.claude.com/docs/en/api/messages/create
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:405]: i am not sure i follow this. are "we" using them as constants? can they actually change?

File: lib/srv/app/llm/anthropic_test.go [hunk 15]
```diff
@@ -0,0 +1,777 @@
...
+	target *url.URL
+}
+
+func (t *redirectingTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	r.URL.Scheme = t.target.Scheme
+	r.URL.Host = t.target.Host
+	return http.DefaultTransport.RoundTrip(r)
+}
+
+func TestReadAnthropicRequest(t *testing.T) {
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic_test.go:718]: what if the input is not valid json?

File: lib/srv/app/llm/handler.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	if c.AWSConfigProvider == nil {
+		var err error
+		c.AWSConfigProvider, err = awsconfig.NewCache()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	if c.HTTPClient == nil {
+		client, err := defaults.HTTPClient()
```
Comment (greedy52) [inline lib/srv/app/llm/handler.go:67]: i assume llm proxy will ignore app-specific tls settings like insecure?

File: lib/srv/app/llm/handler.go [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+		h.handleAnthropic(sessionCtx, w, r)
+		return nil
+	default:
+		return trace.NotImplemented("llm format %q not supported", format)
+	}
+}
+
+// readLimitedRequestBody reads request body respecting the app service memory
+// limits.
+func (h *Handler) readLimitedRequestBody(r *http.Request, maxSize int64) (*bytes.Buffer, error) {
```
Comment (greedy52) [inline lib/srv/app/llm/handler.go:115]:
```suggestion
func readLimitedRequestBody(r *http.Request, maxSize int64) (*bytes.Buffer, error) {
```

File: lib/srv/app/llm/anthropic.go [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+	opts := []option.RequestOption{
+		option.WithHTTPClient(h.cfg.HTTPClient),
+		option.WithJSONSet(modelRequestKey, req.Model),
+		option.WithMaxRetries(0), // Disable retries.
+		option.WithRequestTimeout(req.Timeout),
+		// Ensure the outgoing request URL always contains exactly one /v1 path
+		// prefix. This protects against misconfigured base URLs (e.g. via
+		// ANTHROPIC_BASE_URL env var) that omit the /v1 segment, and covers the
+		// required v1 prefix for Bedrock endpoints.
+		option.WithMiddleware(ensureV1PathPrefix),
```
Comment (greedy52) [inline lib/srv/app/llm/anthropic.go:162]: i wonder what happens if this implementation sees v2 path some day

**Review by greedy52 — COMMENTED** [review]
first pass

File: lib/srv/app/llm/anthropic.go [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+	defer func() {
+		providerRequestDurationHist.WithLabelValues(
+			teleport.ComponentLLM,
+			llm.Format,
+			req.Provider,
+			strconv.FormatBool(req.Streaming),
+		).Observe(time.Since(start).Seconds())
+	}()
+
+	opts := []option.RequestOption{
```
Comment (tigrato) [inline lib/srv/app/llm/anthropic.go:153]: which kind of transformation does the sdk perform?
can we do it ourselves using the forwarder so we can't read everything into memory?
